crates/sui-framework/packages/sui-framework/sources/k_of_n_upgrade_policy.move

```diff
@@ -0,0 +1,395 @@
+            voter_idx = voter_idx - 1;
+            let address = *vector::borrow(voter_addresses, voter_idx);
+            let voter_uid = object::new(ctx);
+            let voter_id = object::uid_to_inner(&voter_uid);
+            transfer::transfer(
+                VotingCap {
+                    id: voter_uid,
+                    owner: address,
+                    upgrade_cap: cap_id,
+                    transfers_count: 0,
+                    votes_issued: 0,
+                },
+                address,
+            );
+            vec_set::insert(&mut voter_caps, voter_id);
+        };
+
+        KofNUpgradeCap {
+            id: cap_uid,
+            upgrade_cap,
+            required_votes,
+            voters,
+            voter_caps,
+        }
+    }
+
+    /// Make the package immutable by destroying the k of n upgrade cap and the
+    /// underlying upgrade cap.
+    public fun make_immutable(cap: KofNUpgradeCap) {
+        let KofNUpgradeCap {
+            id,
+            upgrade_cap,
+            required_votes: _,
+            voters: _,
+            voter_caps: _,
+        } = cap;
+        object::delete(id);
+        package::make_immutable(upgrade_cap);
+    }
+
+    /// Restrict upgrades to "add code only", or "change dependencies".
+    public fun only_additive_upgrades(cap: &mut KofNUpgradeCap) {
+        package::only_additive_upgrades(&mut cap.upgrade_cap)
+    }
+
+    /// Restrict upgrades to "change dependencies only".
+    public fun only_dep_upgrades(cap: &mut KofNUpgradeCap) {
+        package::only_dep_upgrades(&mut cap.upgrade_cap)
+    }
+
+    /// Propose an upgrade. 
+    /// The `digest` of the proposed upgrade is provided to identify the upgrade.
+    /// The proposer is the sender of the transaction and must be the signer
+    /// of the commit transaction as well.
+    public fun propose_upgrade(
+        cap: &KofNUpgradeCap,
+        digest: vector<u8>,
+        ctx: &mut TxContext,
+    ) {
+        let cap_id = object::id(cap);
+        let proposal_uid = object::new(ctx);
+        let proposal_id = object::uid_to_inner(&proposal_uid);
+        
+        let proposer = tx_context::sender(ctx);
+        
+        event::emit(UpgradeProposed {
+            upgrade_cap: cap_id,
+            proposal: proposal_id,
+            digest,
+            proposer,
+            voters: cap.voters,
+        });
+
+        transfer::share_object(ProposedUpgrade {
+            id: proposal_uid,
+            upgrade_cap: cap_id,
+            proposer,
+            digest,
+            current_voters: vec_set::empty(),
+        })
+    }
+
+    /// Vote in favor of an upgrade, aborts if the voter is not for the proposed
+    /// upgrade or if they voted already, or if the upgrade was already performed.
+    public fun vote(
+        proposal: &mut ProposedUpgrade, 
+        voter: &mut VotingCap,
+        ctx: &TxContext,
+    ) {
+        assert!(proposal.proposer != @0x0, EAlreadyIssued);
```

Comment:
Will we have shared object deletion by the time this will go out? If so, we can avoid this pattern and just accept the `ProposedUpgrade` by value when authorizing.

Reply:
likely we will not have it otherwise that was the plan!